### ISSUE TITLE
Release SQL JDBC driver

### ISSUE THREAD

**Issue by Yury-Fridlyand** [opensearch-project/opensearch-build#2692]
### Is your feature request related to a problem? Please describe

JDBC driver was never released on Maven. The one published on [artifacts page](https://opensearch.org/artifacts) is a bit old already and should be updated.

### Describe the solution you'd like

Publish most recent JDBC driver on Maven and on artifacts page.
Driver build is available as GHA artifact: https://github.com/opensearch-project/sql/actions/workflows/sql-jdbc-test-and-build-workflow.yml?query=branch%3A2.x

### Describe alternatives you've considered

N/A

### Additional context

https://github.com/opensearch-project/opensearch-build/issues/794
https://github.com/opensearch-project/sql/issues/401

CC: @anirudha

**Comment by prudhvigodithi** [comment]
Hey @Yury-Fridlyand we have some [publish scripts](https://github.com/opensearch-project/opensearch-build/tree/main/publish) handy, could use them, but quick question, do you have a proper project or gradle task when executed publishes SQL JDBC driver jar to local maven repo (~/.m2) with proper maven group Id like `org.opensearch.driver`? may be if we have that task or job we can use it part of the CI to publish it to actual maven repo with right creds.
Example for [opensearch-job-scheduler](https://repo1.maven.org/maven2/org/opensearch/plugin/opensearch-job-scheduler/2.3.0.0/opensearch-job-scheduler-2.3.0.0.pom) plugin zip
```
<groupId>org.opensearch.plugin</groupId>
<artifactId>opensearch-job-scheduler</artifactId>
<version>2.3.0.0</version>
```
Thank you

**Comment by Yury-Fridlyand** [comment]
Thanks, I will have a look and make all preparations. We need to do few minor fix in the driver prior to release.
I'll ping you once everything will be ready.

**Comment by MaxKsyunz** [comment]
@prudhvigodithi yes there's a `publishToMavenLocal` task in jdbc driver.

One quirk with JDBC driver is that it can build two jars -- normal one, and one with all dependencies shadow copied. The second one is what can be used right now by just dropping the jar in the classpath. We want to publish the first one.

**Comment by MaxKsyunz** [comment]
Notes from our discussion.

Examples to follow:
- opensearch-java publishes with Jenkins like [this](https://github.com/opensearch-project/opensearch-java/blob/main/jenkins/stage-maven-release.jenkinsfile).
- opensearch-testcontainers publishes with Jenkins like [this](https://github.com/opensearch-project/opensearch-testcontainers/blob/main/jenkins/publish-snapshot.jenkinsfile).
JDBC driver will be set up similarly.

Someone with   admin permissions on sql repo will need to setup a webhook used by Jenkins to publish releases when a release tag is made.

**Comment by MaxKsyunz** [comment]
@peterzhuamazon since JDBC driver is in the sql repo, what git tag pattern should its releases have? The Jenkins pipelines currently assume `refs/tags/v`

**Comment by MaxKsyunz** [comment]
Also @peterzhuamazon do you know if JDBC driver will be kept in the plugin repo or split out? The discussion happened [here](https://github.com/opensearch-project/opensearch-build/issues/2188) but it's not clear what will happen to drivers.

If JDBC driver stays in a monorepo with something else, will there be one Jenkinsfile for each product in the repo or one Jenkinsfile for all of repo?

**Comment by peterzhuamazon** [comment]
Hi @MaxKsyunz I added several comments to @Yury-Fridlyand PR there.
Would love to discussion more later on the strategies of the repo.

Thanks.

**Comment by dblock** [comment]
@MaxKsyunz Is there an issue on extracting the JDBC driver into its own repo with pros/cons already? Care to open one so we can discuss that there?

My uneducated opinion is that it looks like the JDBC driver should follow its own semver, and have its own CI/CD that publishes it to maven similar to how we do this now in opensearch-java and other repos. Sounds to me like it should live in its own repo, but I don't know what its dependencies are.

**Comment by acarbonetto** [comment]
> My uneducated opinion is that it looks like the JDBC driver should follow its own semver, and have its own CI/CD that publishes it to maven similar to how we do this now in opensearch-java and other repos. Sounds to me like it should live in its own repo, but I don't know what its dependencies are.0

Raised: 
* https://github.com/opensearch-project/opensearch-build/issues/2799
* https://github.com/opensearch-project/opensearch-build/issues/2800
* https://github.com/opensearch-project/opensearch-build/issues/2801
for specific discussion and execution @dblock - thanks!

**Comment by anirudha** [comment]
is repo split a blocker for maven release now on this issue or is it a separate issue ?

**Comment by acarbonetto** [comment]
> is repo split a blocker for maven release now on this issue or is it a separate issue ?

Separate issue.  At least until I'm informed otherwise.

**Comment by dblock** [comment]
AFAIK if we were able to release the driver before I don't see why splitting should be a blocker, however to setup automation you'll need to queue off a tag, and that becomes more difficult in a monorepo (you'd have to queue of a tag format, unclear which branch to tag, etc.).

**Comment by peterzhuamazon** [comment]
Hi,

Have new discussion with @MaxKsyunz @Yury-Fridlyand @penghuo here is the method we are going to use:

1. Tag creation trigger a github action to clone the code, build maven artifacts, and push to an s3 bucket.
2. Github action will trigger jenkins workflow with the webhook.
3. Jenkins workflow will download the maven artifacts from s3 bucket, sign them, stage to sonatype.
4. Build repo members login to sonatype to release artifacts to maven central.

Thanks.

**Comment by anirudha** [comment]
thanks peter, can you provide n estimate on this.

**Comment by Yury-Fridlyand** [comment]
Release automation is tracked in #2505. This ticket is to track the current JDBC release.
The deadline is ~ Nov 9, 2022.

---------
I created a JDBC driver pre-release on SQL repo: https://github.com/opensearch-project/sql/releases/tag/JDBC-2.0.0.0
Please proceed with manual publish on the Maven ([no-shadow-copy-jar.zip](https://github.com/opensearch-project/sql/files/9932614/no-shadow-copy-jar.zip)).

**Comment by YANG-DB** [comment]
We recently separated SQL into multiple repositories - this issue is now moving to the appropriate repo
https://github.com/opensearch-project/sql-jdbc

**Comment by gaiksaya** [comment]
Hi,
I am looking into integrating the workflows for releasing SQL-JDBC. Publishing snapshots integration is depending on #1182 
For releasing artifacts to maven central, I'll create PR with above repo.
Thanks!

**Comment by gaiksaya** [comment]
Hi,

@MaxKsyunz @YANG-DB @anirudha,

Looks like `./gradlew build test shadowJar` is just generating a single jar. Can you follow the instructions posted here: https://github.com/opensearch-project/opensearch-build/issues/2692#issuecomment-1278043132 and it get it in so that we can publish to maven?
Example: https://aws.oss.sonatype.org/content/repositories/releases/org/opensearch/opensearch-job-scheduler/2.4.1.0/
This has source java doc and application jar under same folder. Ideally it should be this way so that a user can download all 3 of them with same groupID. Also the gradle command should generate it all in a folder org/opensearch/driver. Can someone look into that?
Thanks!

**Comment by YANG-DB** [comment]
can we remove jdbc-stage-maven-release.jenkinsfile

**Comment by gaiksaya** [comment]
The jenkins file will be replaced similar by the ones similar to https://github.com/opensearch-project/spring-data-opensearch/blob/main/jenkins/release.jenkinsFile
For generating release artifacts we need some gradle command similar to this line https://github.com/opensearch-project/spring-data-opensearch/blob/main/.github/workflows/release-drafter.yml#L24 for jbdc that will take care of everything stated here https://github.com/opensearch-project/opensearch-build/issues/2692#issuecomment-1371566132

**Comment by YANG-DB** [comment]
OK
I will replace the jenkins file with a similar to the one you suggested -
can u plz take care for the creation of the appropriate
tokenIdCredential:
*'jenkins-sql-jdbc-opensearch-generic-webhook-token'
*?
Thanks

‫בתאריך יום ד׳, 4 בינו׳ 2023 ב-19:48 מאת ‪Sayali Gaikawad‬‏ <‪
***@***.***‬‏>:‬

**Comment by gaiksaya** [comment]
Hey! Looks like there is some confusion. Sorry for not being clear. I'll take care of adding release workflows (GHA workflow as well as jenkins), also will set up other things like webhook and stuff. However, before we proceed with any of that, jdbc team needs to fix what I mentioned [here](https://github.com/opensearch-project/opensearch-build/issues/2692#issuecomment-1371566132) in above comment.
Please update this issue once its done.
Thanks!

**Comment by gaiksaya** [comment]
Thanks @Yury-Fridlyand , that would be great! So as far I understand this is what we want to publish on maven right?
Got confused because of the shadow jar conversation.

**Comment by Yury-Fridlyand** [comment]
@gaiksaya, please, see https://github.com/opensearch-project/sql-jdbc/pull/45

**Comment by YANG-DB** [comment]
@gaiksaya - plz add the 'release_drafter.yml' workflow to the following branches 
 - main
 - 2.x
Thanks

**Comment by gaiksaya** [comment]
Hey @YANG-DB ,
Github actions only triggers workflows on default branch (which I assume is main). The triggers are at repo level. So it you want to publish something on 2.x branch, just checkout the branch, create a tag and push it. It will trigger the workflow and checkout the git ref accordingly. 
Will work on adding the release workflows. Thanks!

**Comment by gaiksaya** [comment]
Few questions:
1. What should be the trigger for the release? The ones followed by various other repos in this org is using a tag. However, I had offline discussion with @YANG-DB that they want to control publishing snapshots (in future) via tag too. 
2. Please confirm the below artifacts needs to be published to maven:
```
build/repository
└── org
    └── opensearch
        └── driver
            └── opensearch-sql-jdbc
                ├── 1.2.0.0
                │   ├── opensearch-sql-jdbc-1.2.0.0.jar
                │   ├── opensearch-sql-jdbc-1.2.0.0.jar.md5
                │   ├── opensearch-sql-jdbc-1.2.0.0.jar.sha1
                │   ├── opensearch-sql-jdbc-1.2.0.0.pom
                │   ├── opensearch-sql-jdbc-1.2.0.0.pom.md5
                │   └── opensearch-sql-jdbc-1.2.0.0.pom.sha1
                ├── maven-metadata.xml
                ├── maven-metadata.xml.md5
                └── maven-metadata.xml.sha1
```

**Comment by YANG-DB** [comment]
hi
1. tag is fine
2. yes @Yury-Fridlyand can you approve ?

**Comment by gaiksaya** [comment]
Thanks,
Please make sure you do not use tag for snapshots as well then as it will draft a release, triggering the release workflows. For  publishing snapshots I believe `push` is a good way to go. Your call. 

One more question:
What should be the java version? Need it to specify here https://github.com/opensearch-project/spring-data-opensearch/blob/main/.github/workflows/release-drafter.yml#L19

**Comment by YANG-DB** [comment]
lets go with java 11
@Yury-Fridlyand - sound correct ?

**Comment by Yury-Fridlyand** [comment]
> lets go with java 11
> @Yury-Fridlyand - sound correct ?

Yes. Small note - version should be 2.0.0.0 or higher.
Please, check that you are on the right repo (`sql-jdbc`, not `sql`) and right branch.

**Comment by gaiksaya** [comment]
Hi @YANG-DB , @joshuali925 , @anirudha 
Was the security review completed for this version? Since the [release template](https://github.com/opensearch-project/opensearch-build/blob/main/.github/ISSUE_TEMPLATE/standalone_releases_template.yaml) was not used, asking this question here.
Please confirm. Thanks!
cc: @bbarani

**Comment by gaiksaya** [comment]
Got confirmation from @dblock @bbarani that security review is not required since its not a new product just code shifted to new repo. All the set up is done. We are good for the release. Will sync up with @YANG-DB offline to cut a tag.
Thanks!

**Comment by gaiksaya** [comment]
sql-jdbc 1.2.0.0 is out https://aws.oss.sonatype.org/content/repositories/releases/org/opensearch/driver/opensearch-sql-jdbc/1.2.0.0/
Thanks everyone involved in this release. The process is end to end automated. You can check more details about universal release mechanism in this issue https://github.com/opensearch-project/opensearch-build/issues/1234.
Regarding releases see the documentation in the repo https://github.com/opensearch-project/sql-jdbc/blob/main/RELEASING.md

**Comment by YANG-DB** [comment]
The SQL [2.5 release](https://github.com/opensearch-project/sql/tree/2.5),  [2.x](https://github.com/opensearch-project/sql/tree/2.x) and [main](https://github.com/opensearch-project/sql/tree/main) branches are now depended on this (maven artifact )[https://repo1.maven.org/maven2/org/opensearch/driver/opensearch-sql-jdbc/1.2.0.0/]

**Comment by gaiksaya** [comment]
Hi @YANG-DB, @joshuali925 

You can now publish snaphots to maven using GHA directly. See the PR[1] as an example on what all needs to be added to make that happen. You can add me as a reviewer once PR is ready so that I can make changes at the backend and add the required role to retrieve the credentials.
Thanks!

[1] https://github.com/opensearch-project/opensearch-java/pull/325
